### PR TITLE
Add encode_varule_to_box(), clean up organization of `custom` module

### PR DESCRIPTION
**Pull request by Manishearth**
Progress towards https://github.com/unicode-org/icu4x/pull/1609 and https://github.com/unicode-org/icu4x/issues/1190

cc @iainireland since you wanted this (though soon enough we'll have a proc macro handling all of this, #1609 has already landed!)

### CODE REVIEW

File: utils/zerovec/src/ule/custom.rs [hunk 1]
```diff
@@ -2,36 +2,43 @@
...
 //!    field1: char,
 //!    field2: u32,
+//!    #[serde(borrow)]
 //!    field3: ZeroVec<'a, u32>   
 //! }
 //! ```
 //!
-//! then the ULE type will be implemented as follows:
+//! then the ULE type will be implemented as follows. Ideally, you should have
+//! and `EncodeAsVarULE` and `ZeroFrom` implementations on `Foo` pertaining to `FooULE`,
```
Comment (iainireland) [inline utils/zerovec/src/ule/custom.rs:27]: Driveby: "you should have and" -> "you should have"?

File: utils/zerovec/src/ule/encode.rs [hunk 2]
```diff
@@ -76,6 +77,24 @@ pub unsafe trait EncodeAsVarULE<T: VarULE + ?Sized> {
...
+    vec.resize(x.encode_var_ule_len(), 0);
+    x.encode_var_ule_write(&mut vec);
+    unsafe {
+        // safety: vec is known to contain a valid encoded T
+        let ptr: *mut T = T::from_byte_slice_unchecked(&vec) as *const T as *mut T;
+        mem::forget(vec);
+
+        // Safety: we can construct an owned version since we have mem::forgotten the older owner
+        Box::from_raw(ptr)
+    }
```
Comment (sffc) [inline utils/zerovec/src/ule/encode.rs:95]: Issue: The memory layout of a `Vec` might not be the same as of a `Box`, I think.

```suggestion
    let boxed_bytes = vec.into_boxed_slice();
    unsafe {
        // safety: boxed_bytes is known to contain a valid encoded T
        let ptr: *mut T = T::from_byte_slice_unchecked(&boxed_bytes) as *const T as *mut T;
        mem::forget(boxed_bytes);

        // Safety: `ptr` is a box, and `T` is a VarULE which guarantees it has the same memory layout as `[u8]`
        Box::from_raw(ptr)
    }
```

Do we even need to call `from_byte_slice_unchecked`?  Can we get the box and convert the pointer type in the box?

**Review by sffc — CHANGES_REQUESTED** [review]

File: utils/zerovec/src/ule/encode.rs (hunk 2 above)
Comment (sffc) [reply utils/zerovec/src/ule/encode.rs:95]: What I mean is, I think it's not sound to convert `Vec<u8>` into `Box<T>`.  We should instead convert `Box<[u8]>` into `Box<T>`.  And, if we do that, then can we cast the pointer type directly, since the length information is encoded into the header of the slice?  We don't really need `from_byte_slice_unchecked` I think.

Comment (sffc) [reply utils/zerovec/src/ule/encode.rs:95]: We have `from_byte_slice_unchecked` because supposedly we couldn't write a general implementation of `&[u8]` into `&T`, but I don't remember why not.

Comment (Manishearth) [reply utils/zerovec/src/ule/encode.rs:95]: No, an allocated `Vec` is guaranteed to be backed by a `Box`able allocation. I can do an explicit `into_boxed_vec()` step there if you'd like, but `Box<[u8]>` to `Box<T>` is basically equivalent to go from `Vec<u8>` to `Box<T>`.

>  And, if we do that, then can we cast the pointer type directly, since the length information is encoded into the header of the slice? 
> We don't really need `from_byte_slice_unchecked` I think.

We cannot do this, the length metadata is not generically know. This is why `from_byte_slice_unchecked()` exists.

DST pointer metadata is not necessarily the length of the allocation, it's basically only that for `str` and `[T]`, not custom DSTs

Comment (sffc) [reply utils/zerovec/src/ule/encode.rs]: In addition to `.into_boxed_slice()` I had also suggested updating the safety comment.

```
// Safety: `ptr` is a box, and `T` is a VarULE which guarantees it has the same memory layout as `[u8]`
```

**Review by sffc — CHANGES_REQUESTED** [review]
.

**Review by sffc — APPROVED** [review]
